src/hardware/serialport/serialport.cpp

```diff
@@ -93,108 +95,86 @@ device_COM::~device_COM() {
 
 
 // COM1 - COM4 objects
-CSerial* serialports[4] ={0,0,0,0};
+CSerial *serialports[SERIAL_MAX_PORTS] = {nullptr};
 
 static Bitu SERIAL_Read (Bitu port, Bitu iolen) {
-	Bitu i;
-	Bitu retval;
-	Bitu index = port & 0x7;
-	switch(port&0xff8) {
-		case 0x3f8: i=0; break;
-		case 0x2f8: i=1; break;
-		case 0x3e8: i=2; break;
-		case 0x2e8: i=3; break;
-		default: return 0xff;
+	(void)iolen; // unused, but required for API compliance
+	uint32_t i = 0;
+	uint32_t retval = 0;
+	uint8_t offset_type = static_cast<uint8_t>(port) & 0x7;
+	switch (static_cast<uint32_t>(port) & 0xff8) {
+	case 0x3f8: i = 0; break;
+	case 0x2f8: i = 1; break;
+	case 0x3e8: i = 2; break;
+	case 0x2e8: i = 3; break;
+	default: return 0xff;
 	}
-	if(serialports[i]==0) return 0xff;
-
-	switch (index) {
-		case RHR_OFFSET:
-			retval = serialports[i]->Read_RHR();
-			break;
-		case IER_OFFSET:
-			retval = serialports[i]->Read_IER();
-			break;
-		case ISR_OFFSET:
-			retval = serialports[i]->Read_ISR();
-			break;
-		case LCR_OFFSET:
-			retval = serialports[i]->Read_LCR();
-			break;
-		case MCR_OFFSET:
-			retval = serialports[i]->Read_MCR();
-			break;
-		case LSR_OFFSET:
-			retval = serialports[i]->Read_LSR();
-			break;
-		case MSR_OFFSET:
-			retval = serialports[i]->Read_MSR();
-			break;
-		case SPR_OFFSET:
-			retval = serialports[i]->Read_SPR();
-			break;
+	if (!serialports[i])
+		return 0xff;
+
+	switch (offset_type) {
+	case RHR_OFFSET: retval = serialports[i]->Read_RHR(); break;
+	case IER_OFFSET: retval = serialports[i]->Read_IER(); break;
+	case ISR_OFFSET: retval = serialports[i]->Read_ISR(); break;
+	case LCR_OFFSET: retval = serialports[i]->Read_LCR(); break;
+	case MCR_OFFSET: retval = serialports[i]->Read_MCR(); break;
+	case LSR_OFFSET: retval = serialports[i]->Read_LSR(); break;
+	case MSR_OFFSET: retval = serialports[i]->Read_MSR(); break;
+	case SPR_OFFSET: retval = serialports[i]->Read_SPR(); break;
 	}
 
 #if SERIAL_DEBUG
 	const char* const dbgtext[]=
 		{"RHR","IER","ISR","LCR","MCR","LSR","MSR","SPR","DLL","DLM"};
 	if(serialports[i]->dbg_register) {
-		if((index<2) && ((serialports[i]->LCR)&LCR_DIVISOR_Enable_MASK))
-			index += 8;
+		if ((offset_type < 2) &&
+		    ((serialports[i]->LCR) & LCR_DIVISOR_Enable_MASK))
+			offset_type += 8;
 		serialports[i]->log_ser(serialports[i]->dbg_register,
-			"read  0x%2x from %s.",retval,dbgtext[index]);
+		                        "read  0x%2x from %s.", retval,
+		                        dbgtext[offset_type]);
 	}
 #endif
-	return retval;	
+	return static_cast<Bitu>(retval);
```

Comment:
This function still returns a `Bitu` to complete with a wide-reaching DOSBox API; so we're stuck with this here for now.